onnx/defs/gen_doc.py

```diff
@@ -139,13 +133,108 @@ def main(args):
+            json_schema['doc'] = schema.doc.lstrip();
+        if schema.inputs:
+            json_schema['inputs'] = []
+            for input in schema.inputs:
+                json_schema['inputs'].append({ 
+                    'name': input.name, 
+                    'description': input.description,
+                    'optional': input.optional,
+                    'typeStr': input.typeStr,
+                    'types': generate_json_types(input.types) })
+        json_schema['min_input'] = schema.min_input;
+        json_schema['max_input'] = schema.max_input;
+        if schema.outputs:
+            json_schema['outputs'] = []
+            for output in schema.outputs:
+                json_schema['outputs'].append({ 
+                    'name': output.name, 
+                    'description': output.description,
+                    'optional': output.optional,
+                    'typeStr': output.typeStr,
+                    'types': generate_json_types(output.types) })
+        json_schema['min_output'] = schema.min_output;
+        json_schema['max_output'] = schema.max_output;
+        if schema.attributes:
+            json_schema['attributes'] = []
+            for _, attribute in sorted(schema.attributes.items()):
+                json_schema['attributes'].append({
+                    'name' : attribute.name,
+                    'description': attribute.description,
+                    'type': generate_json_attr_type(attribute.type),
+                    'required': attribute.required })
+        if schema.type_constraints:
+            json_schema["type_constraints"] = []
+            for type_constraint in schema.type_constraints:
+                json_schema['type_constraints'].append({
+                    'description': type_constraint.description,
+                    'type_param_str': type_constraint.type_param_str,
+                    'allowed_type_strs': type_constraint.allowed_type_strs
+                })
+        if op_type in SNIPPETS:
+            json_schema['snippets'] = []
+            for summary, code in sorted(SNIPPETS[op_type]):
+                json_schema['snippets'].append({
+                    'summary': summary,
+                    'code': code
+                })
+        json_root.append({
+            "op_type": op_type,
+            "schema": json_schema })
+    with io.open(file, 'w', newline='') as fout:
+        json_root = json.dumps(json_root, sort_keys=True, indent=2)
+        for line in json_root.splitlines():
+            line = line.rstrip()
+            if sys.version_info[0] < 3:
+                line = unicode(line)
+            fout.write(line)
+            fout.write('\n')
+
+def main(args):
+    sorted_ops = sorted(
+        (int(schema.support_level), op_type, schema)
+        for (op_type, schema) in defs.get_all_schemas().items())
+    # If no format is specified, generate /docs/Operator.md and /docs/Operator.json by default
+    if args.format == '':
```

Comment:
May I know the point/usage of generating operator.json please? Do you want to check in this operator.json like operator.md? I think we don't want to check it in, if that's the case, then please set the default format running without generating operator.json.

Reply:
To goal is to generate and check-in both. Operator.md is human readable, Operator.json is machine readable. There is no way to get this information in a machine readable form without installing and running the onnx Python code otherwise. The .proto file and this info would be needed to write tools that depend on the data in a model.db but do not run the actual onnx Python code.

Reply:
Sorry, what information do you need from this .json doc please? I think you may call OperatorSchemaRegistry to access all those information programatically instead of parsing a json doc. We have both python and c++ API available there.

Reply:
Those APIs only work if the onnx packages are built and installed on a machine. If onnx is not installed these APIs are not available. The goal is to have some machine processable "schema" that can be used by tools to annotate a model.db graph with more information without requiring onnx installed. This file could be generated on a dev machine but ideally onnx.proto and Operator.json should represent a consistent snapshot for each commit and gen_doc.py seems to be the current approach to make sure they are.

Reply:
Thank you for detail clarification. Hmmm, I personally have some concern of having this json (maintained) checked in as another representation of ONNX "standard". I'd hear some other voices.

Reply:
Depends on what kind of tooling you're thinking about. It's important to make clear that we don't want backend implementations to consume this file and instead they should depend on onnx lib itself if possible. But if you need to build some javascript functionality (e.g. Atom plugin) it might be a reasonable approach.

Putting it in repo vs generating on-the-fly doesn't matter that much. If you're putting it in repo - do you plan to package it in the pip distribution too?

Reply:
Tooling I was thinking about is of the javascript kind. Based on the thread @gramalingam mentioned below it sounds like the right approach is to abandon this PR for now. I can build the JSON file locally as a stop gap and wait for proper support via OperatorProto. Ideally the OperatorProto file should be serialized to JSON as well (to be readable) and get checked in to a known location but getting the proto definitions right probably requires some coordination.